File: py/guitar.py
class Guitar:
    def __init__(self,numberOfStrings=6,tuning="standard",numberOfFrets=19):
        self.numToNoteDict = {
                                0:"A",
                                1:"A#/Bb",
                                2:"B",
                                3:"C",
                                4:"Db/C#",
                                5:"D",
                                6:"Eb/D#",
                                7:"E",
                                8:"F",
                                9:"Gb/F#",
                                10:"G",
                                11:"Ab/G#"
                              }
        
        self.TuneGuitar(numberOfStrings,tuning,numberOfFrets)
    
    def TuneGuitar(self,numberOfStrings=6,tuning="standard",numberOfFrets=19):
        self.tuning = tuning
        if numberOfStrings == 6:
            if tuning == "standard":
                self.stringNotes = [7,2,10,5,0,7]
            elif tuning == "drop d":
                self.stringNotes = [7,2,10,5,0,5]
            elif tuning == "dadgad":
                self.stringNotes = [5,0,10,5,0,5]
            elif tuning == "double drop d":
                self.stringNotes = [5,2,10,5,0,5]
            elif tuning == "open d":
                self.stringNotes = [5,0,9,5,0,5]
            elif tuning == "open e":
                self.stringNotes = [7,2,11,7,2,7]
            elif tuning == "open g":
                self.stringNotes = [5,2,10,5,10,5]
            elif tuning == "open a":
                self.stringNotes = [7,4,0,7,0,7]
            elif tuning == "dgcgcd":
                self.stringNotes = [5,3,10,3,10,5]
            elif tuning == "open c6":
                self.stringNotes = [7,3,10,3,0,3]
        elif numberOfStrings == 12:
            if tuning == "standard":
                self.stringNotes = [7,7,2,2,10,10,5,5,0,0,7,7]
            
        self.CreateGuitar(numberOfStrings,numberOfFrets)
    
    def CreateGuitar(self,numberOfStrings=6,numberOfFrets=19): # Called by itself if tuning not changed
        self.guitar = []
        for stringNum in range(numberOfStrings):
            noteNum = self.stringNotes[stringNum]
            leadNoteString = self.numToNoteDict[noteNum]
            self.guitar.append([((stringNum,0),(noteNum,leadNoteString))])
        for stringNum in range(numberOfStrings):
            for fretNum in range(1,numberOfFrets+1):
                openNoteNum = self.guitar[stringNum][0][1][0]
                fretNoteNum = (openNoteNum+fretNum) % 12
                neckNoteString = self.numToNoteDict[fretNoteNum]
                self.guitar[stringNum].append(((stringNum,fretNum),(fretNoteNum,neckNoteString)))

File: js/guitar.js
"use strict";

class Guitar {
    constructor({numberOfStrings=6, tuning="standard", numberOfFrets=20} = {}) {
        this.numToNoteDict = [ "A","Bb/A#","B","C","Db/C#","D","Eb/D#","E","F","Gb/F#","G","Ab/G#" ];

        this._tuneGuitar({numberOfStrings: numberOfStrings, tuning: tuning, numberOfFrets: numberOfFrets});
    }

    _tuneGuitar({numberOfStrings=6, tuning="standard", numberOfFrets=20}) {
        this.tuning = tuning;
        if (numberOfStrings === 6) {
            if (tuning === "standard") {
                this.stringNotes = [7,2,10,5,0,7];
            }
            else if (tuning === "drop d") {
                this.stringNotes = [7,2,10,5,0,5];
            }
            else if (tuning === "dadgad") {
                this.stringNotes = [5,0,10,5,0,5];
            }
            else if (tuning === "double drop d") {
                this.stringNotes = [5,2,10,5,0,5];
            }
            else if (tuning === "open d") {
                this.stringNotes = [5,0,9,5,0,5];
            }
            else if (tuning === "open e") {
                this.stringNotes = [7,2,11,7,2,7];
            }
            else if (tuning === "open g") {
                this.stringNotes = [5,2,10,5,10,5];
            }
            else if (tuning === "open a") {
                this.stringNotes = [7,4,0,7,0,7];
            }
            else if (tuning === "dgcgcd") {
                this.stringNotes = [5,3,10,3,10,5];
            }
            else if (tuning === "open c6") {
                this.stringNotes = [7,3,10,3,0,3];
            }
        }
        else if (numberOfStrings === 12) {
            if (tuning === "standard") {
                this.stringNotes = [7,7,2,2,10,10,5,5,0,0,7,7];
            }
        }
            
        this._createGuitar({numberOfStrings: numberOfStrings, numberOfFrets: numberOfFrets});
    }

    _createGuitar({numberOfStrings=6, numberOfFrets=24}) {
    //Called by itself if tuning not changed
        this.guitar = [];
        this.numberOfFrets = numberOfFrets
        let noteNum;
        let leadNoteString;
        // Add open string information to guitar
        for (let stringNum = 0; stringNum < numberOfStrings; stringNum++) {
            noteNum = this.stringNotes[stringNum];
            leadNoteString = this.numToNoteDict[noteNum];
            this.guitar.push([[[stringNum,0],[noteNum,leadNoteString]]]);
        }

        let openNoteNum;
        let fretNoteNum;
        let neckNoteString;
        // Add remaining string information
        for (let stringNum = 0; stringNum < numberOfStrings; stringNum++) {
            for (let fretNum = 1; fretNum < numberOfFrets+1; fretNum++) {
                openNoteNum = this.guitar[stringNum][0][1][0];
                fretNoteNum = (openNoteNum+fretNum) % 12;
                neckNoteString = this.numToNoteDict[fretNoteNum];
                this.guitar[stringNum].push([[stringNum,fretNum],
                                            [fretNoteNum,neckNoteString]]);
            }
        }
    }
}